spec/support/routes_and_redirects_validator.rb

```diff
@@ -0,0 +1,226 @@
+RSpec.shared_examples_for RoutesAndRedirectsValidator do
+  describe "routes validations" do
+    it "is invalid when a route is not below the base path" do
+      subject.routes = [
+        { path: subject.base_path, type: "exact" },
+        { path: "/wrong-path", type: "exact" },
+      ]
+
+      expect(subject).to be_invalid
+      expect(subject.errors[:routes]).to eq(["path must be below the base path"])
+    end
+
+    it "must have unique paths" do
+      subject.routes = [
+        { path: subject.base_path, type: "exact" },
+        { path: subject.base_path, type: "exact" },
+      ]
+
+      expect(subject).to be_invalid
+      expect(subject.errors[:routes]).to eq(["must have unique paths"])
+    end
+
+    it "must have a type" do
+      subject.routes = [{ path: subject.base_path }]
+      expect(subject).to be_invalid
+    end
+
+    it "must have a path" do
+      subject.routes = [{ type: "exact" }]
+      expect(subject).to be_invalid
+    end
+
+    it "must have a valid type" do
+      subject.routes = [{ path: subject.base_path, type: "unsupported" }]
+      expect(subject).to be_invalid
+    end
+
+    it "cannot have extra keys" do
+      subject.routes = [{ path: subject.base_path, type: "exact", foo: "bar" }]
+
+      expect(subject).to be_invalid
+      expect(subject.errors[:routes]).to eq(["unsupported keys: foo"])
+    end
+
+    it "is valid with a dashed locale" do
+      subject.routes = [
+        { path: subject.base_path, type: "exact" },
+        { path: "#{subject.base_path}.es-419", type: "exact" },
+      ]
+
+      expect(subject).to be_valid
+    end
+
+    it "must contain valid absolute paths" do
+      subject.routes = [
+        { path: subject.base_path, type: "exact" },
+        { path: "#{subject.base_path}/ not valid", type: "exact" },
+      ]
+
+      expect(subject).to be_invalid
+      expect(subject.errors[:routes]).to eq(["is not a valid absolute URL path"])
+    end
+
+    it "does not throw an error when routes is nil rather than an empty array" do
+      subject.routes = nil
+      expect { subject.valid? }.to_not raise_error
+    end
+
+    context "for a redirect item" do
```

Comment:
I don't think this is quite correct.  All formats can include redirects along with the routes. redirect items are special in that they only include redirects (and therefore require the redirects to include the `base_path` instead of the routes).